File: NearMe/Components/CardComponents.swift
import SwiftUI

// MARK: - Card Components
struct BaseCard<Content: View>: View {
    let content: Content
    var backgroundColor: Color = DesignSystem.Colors.card
    var cornerRadius: CGFloat = DesignSystem.CornerRadius.md
    var shadowStyle: ShadowStyle = DesignSystem.Shadow.small
    var padding: CGFloat = DesignSystem.Spacing.md
    
    init(backgroundColor: Color = DesignSystem.Colors.card,
         cornerRadius: CGFloat = DesignSystem.CornerRadius.md,
         shadowStyle: ShadowStyle = DesignSystem.Shadow.small,
         padding: CGFloat = DesignSystem.Spacing.md,
         @ViewBuilder content: () -> Content) {
        self.backgroundColor = backgroundColor
        self.cornerRadius = cornerRadius
        self.shadowStyle = shadowStyle
        self.padding = padding
        self.content = content()
    }
    
    var body: some View {
        content
            .padding(padding)
            .background(
                RoundedRectangle(cornerRadius: cornerRadius)
                    .fill(backgroundColor)
            )
            .designSystemShadow(shadowStyle)
    }
}

struct TaskCard: View {
    let title: String
    let description: String?
    let location: String
    let status: TaskStatus
    let action: () -> Void
    
    enum TaskStatus {
        case active, completed, muted
        
        var color: Color {
            switch self {
            case .active: return DesignSystem.Colors.primary
            case .completed: return DesignSystem.Colors.success
            case .muted: return DesignSystem.Colors.textTertiary
            }
        }
        
        var icon: String {
            switch self {
            case .active: return "circle"
            case .completed: return "checkmark.circle.fill"
            case .muted: return "minus.circle.fill"
            }
        }
    }
    
    var body: some View {
        BaseCard {
            VStack(alignment: .leading, spacing: DesignSystem.Spacing.sm) {
                HStack {
                    Image(systemName: status.icon)
                        .foregroundColor(status.color)
                        .font(.title3)
                    
                    VStack(alignment: .leading, spacing: 2) {
                        Text(title)
                            .font(DesignSystem.Typography.title3)
                            .foregroundColor(DesignSystem.Colors.textPrimary)
                            .lineLimit(2)
                        
                        if let description = description {
                            Text(description)
                                .font(DesignSystem.Typography.body)
                                .foregroundColor(DesignSystem.Colors.textSecondary)
                                .lineLimit(3)
                        }
                    }
                    
                    Spacer()
                    
                    IconButton(icon: "chevron.right", action: action, size: 32)
                }
                
                HStack {
                    Image(systemName: "location")
                        .foregroundColor(DesignSystem.Colors.textSecondary)
                        .font(.caption)
                    
                    Text(location)
                        .font(DesignSystem.Typography.caption1)
                        .foregroundColor(DesignSystem.Colors.textSecondary)
                        .lineLimit(1)
                    
                    Spacer()
                }
            }
        }
    }
}

struct PlaceCard: View {
    let name: String
    let address: String
    let category: String?
    let distance: String?
    let action: () -> Void
    
    var body: some View {
        BaseCard {
            VStack(alignment: .leading, spacing: DesignSystem.Spacing.sm) {
                HStack {
                    VStack(alignment: .leading, spacing: 2) {
                        Text(name)
                            .font(DesignSystem.Typography.title3)
                            .foregroundColor(DesignSystem.Colors.textPrimary)
                            .lineLimit(2)
                        
                        Text(address)
                            .font(DesignSystem.Typography.body)
                            .foregroundColor(DesignSystem.Colors.textSecondary)
                            .lineLimit(2)
                    }
                    
                    Spacer()
                    
                    IconButton(icon: "chevron.right", action: action, size: 32)
                }
                
                HStack {
                    if let category = category {
                        Text(category.uppercased())
                            .font(DesignSystem.Typography.caption2)
                            .fontWeight(.medium)
                            .foregroundColor(DesignSystem.Colors.primary)
                            .padding(.horizontal, DesignSystem.Spacing.sm)
                            .padding(.vertical, DesignSystem.Spacing.xs)
                            .background(
                                Capsule()
                                    .fill(DesignSystem.Colors.primary.opacity(0.1))
                            )
                    }
                    
                    if let distance = distance {
                        Text(distance)
                            .font(DesignSystem.Typography.caption1)
                            .foregroundColor(DesignSystem.Colors.textSecondary)
                    }
                    
                    Spacer()
                }
            }
        }
    }
}

struct NotificationCard: View {
    let title: String
    let message: String
    let timestamp: Date
    let isRead: Bool
    let actions: [NotificationAction]
    
    struct NotificationAction {
        let title: String
        let action: () -> Void
        let style: ActionStyle
        
        enum ActionStyle {
            case primary, secondary, destructive
        }
    }
    
    var body: some View {
        BaseCard(
            backgroundColor: isRead ? DesignSystem.Colors.card : DesignSystem.Colors.surface,
            shadowStyle: isRead ? DesignSystem.Shadow.small : DesignSystem.Shadow.medium
        ) {
            VStack(alignment: .leading, spacing: DesignSystem.Spacing.sm) {
                HStack {
                    VStack(alignment: .leading, spacing: 2) {
                        Text(title)
                            .font(DesignSystem.Typography.title3)
                            .foregroundColor(DesignSystem.Colors.textPrimary)
                            .fontWeight(isRead ? .regular : .semibold)
                        
                        Text(message)
                            .font(DesignSystem.Typography.body)
                            .foregroundColor(DesignSystem.Colors.textSecondary)
                            .lineLimit(3)
                    }
                    
                    Spacer()
                    
                    if !isRead {
                        Circle()
                            .fill(DesignSystem.Colors.primary)
                            .frame(width: 8, height: 8)
                    }
                }
                
                HStack {
                    Text(timestamp, style: .relative)
                        .font(DesignSystem.Typography.caption1)
                        .foregroundColor(DesignSystem.Colors.textTertiary)
                    
                    Spacer()
                }
                
                if !actions.isEmpty {
                    HStack(spacing: DesignSystem.Spacing.sm) {
                        ForEach(actions.indices, id: \.self) { index in
                            let action = actions[index]
                            Button(action.title, action: action.action)
                                .font(DesignSystem.Typography.caption1)
                                .fontWeight(.medium)
                                .padding(.horizontal, DesignSystem.Spacing.sm)
                                .padding(.vertical, DesignSystem.Spacing.xs)
                                .background(
                                    RoundedRectangle(cornerRadius: DesignSystem.CornerRadius.sm)
                                        .fill(actionStyleColor(action.style))
                                )
                                .foregroundColor(actionStyleForegroundColor(action.style))
                        }
                        
                        Spacer()
                    }
                }
            }
        }
    }
    
    private func actionStyleColor(_ style: NotificationCard.NotificationAction.ActionStyle) -> Color {
        switch style {
        case .primary: return DesignSystem.Colors.primary
        case .secondary: return DesignSystem.Colors.surface
        case .destructive: return DesignSystem.Colors.error
        }
    }
    
    private func actionStyleForegroundColor(_ style: NotificationCard.NotificationAction.ActionStyle) -> Color {
        switch style {
        case .primary: return DesignSystem.Colors.textInverse
        case .secondary: return DesignSystem.Colors.textPrimary
        case .destructive: return DesignSystem.Colors.textInverse
        }
    }
}

struct EmptyStateCard: View {
    let icon: String
    let title: String
    let message: String
    let actionTitle: String?
    let action: (() -> Void)?
    
    var body: some View {
        BaseCard {
            VStack(spacing: DesignSystem.Spacing.lg) {
                Image(systemName: icon)
                    .font(.system(size: 48, weight: .light))
                    .foregroundColor(DesignSystem.Colors.textTertiary)
                
                VStack(spacing: DesignSystem.Spacing.sm) {
                    Text(title)
                        .font(DesignSystem.Typography.title2)
                        .foregroundColor(DesignSystem.Colors.textPrimary)
                        .multilineTextAlignment(.center)
                    
                    Text(message)
                        .font(DesignSystem.Typography.body)
                        .foregroundColor(DesignSystem.Colors.textSecondary)
                        .multilineTextAlignment(.center)
                }
                
                if let actionTitle = actionTitle, let action = action {
                    PrimaryButton(title: actionTitle, action: action)
                }
            }
            .padding(DesignSystem.Spacing.lg)
        }
    }
}

// MARK: - Card Previews
struct CardComponents_Previews: PreviewProvider {
    static var previews: some View {
        ScrollView {
            VStack(spacing: DesignSystem.Spacing.md) {
                TaskCard(
                    title: "Buy groceries",
                    description: "Pick up milk, bread, and eggs from the store",
                    location: "Whole Foods Market",
                    status: .active,
                    action: {}
                )
                
                PlaceCard(
                    name: "Central Park",
                    address: "New York, NY 10024",
                    category: "Park",
                    distance: "0.5 mi",
                    action: {}
                )
                
                NotificationCard(
                    title: "Task Reminder",
                    message: "You're near Whole Foods Market. Don't forget to buy groceries!",
                    timestamp: Date().addingTimeInterval(-300),
                    isRead: false,
                    actions: [
                        NotificationCard.NotificationAction(
                            title: "Complete",
                            action: {},
                            style: .primary
                        ),
                        NotificationCard.NotificationAction(
                            title: "Snooze",
                            action: {},
                            style: .secondary
                        )
                    ]
                )
                
                EmptyStateCard(
                    icon: "list.bullet",
                    title: "No Tasks Yet",
                    message: "Create your first task to get started with location-based reminders.",
                    actionTitle: "Create Task",
                    action: {}
                )
            }
            .padding()
        }
        .background(DesignSystem.Colors.background)
    }
}

File: android/app/src/main/java/com/nearme/app/ui/components/CardComponents.kt
package com.nearme.app.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.nearme.app.ui.theme.*

@Composable
fun BaseCard(
    modifier: Modifier = Modifier,
    backgroundColor: Color = Card,
    cornerRadius: androidx.compose.ui.unit.Dp = DesignSystem.CornerRadius.md,
    elevation: androidx.compose.ui.unit.Dp = DesignSystem.Elevation.small,
    content: @Composable () -> Unit
) {
    Card(
        modifier = modifier,
        colors = CardDefaults.cardColors(containerColor = backgroundColor),
        shape = RoundedCornerShape(cornerRadius),
        elevation = CardDefaults.cardElevation(defaultElevation = elevation)
    ) {
        Box(
            modifier = Modifier.padding(DesignSystem.Spacing.md)
        ) {
            content()
        }
    }
}

@Composable
fun TaskCard(
    title: String,
    description: String?,
    location: String,
    status: TaskStatus,
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    BaseCard(
        modifier = modifier.clickable { onClick() }
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.sm)
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.md)
            ) {
                Icon(
                    imageVector = androidx.compose.material.icons.Icons.Default.Circle,
                    contentDescription = null,
                    tint = status.color,
                    modifier = Modifier.size(DesignSystem.IconSize.medium)
                )
                
                Column(
                    modifier = Modifier.weight(1f),
                    verticalArrangement = Arrangement.spacedBy(2.dp)
                ) {
                    Text(
                        text = title,
                        style = MaterialTheme.typography.titleMedium,
                        color = TextPrimary,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis
                    )
                    
                    if (description != null) {
                        Text(
                            text = description,
                            style = MaterialTheme.typography.bodyMedium,
                            color = TextSecondary,
                            maxLines = 3,
                            overflow = TextOverflow.Ellipsis
                        )
                    }
                }
                
                Icon(
                    imageVector = androidx.compose.material.icons.Icons.Default.ChevronRight,
                    contentDescription = null,
                    tint = TextTertiary,
                    modifier = Modifier.size(DesignSystem.IconSize.small)
                )
            }
            
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.xs)
            ) {
                Icon(
                    imageVector = androidx.compose.material.icons.Icons.Default.LocationOn,
                    contentDescription = null,
                    tint = TextSecondary,
                    modifier = Modifier.size(DesignSystem.IconSize.small)
                )
                
                Text(
                    text = location,
                    style = MaterialTheme.typography.labelMedium,
                    color = TextSecondary,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis
                )
            }
        }
    }
}

@Composable
fun PlaceCard(
    name: String,
    address: String,
    category: String?,
    distance: String?,
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    BaseCard(
        modifier = modifier.clickable { onClick() }
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.sm)
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.md)
            ) {
                Column(
                    modifier = Modifier.weight(1f),
                    verticalArrangement = Arrangement.spacedBy(2.dp)
                ) {
                    Text(
                        text = name,
                        style = MaterialTheme.typography.titleMedium,
                        color = TextPrimary,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis
                    )
                    
                    Text(
                        text = address,
                        style = MaterialTheme.typography.bodyMedium,
                        color = TextSecondary,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis
                    )
                }
                
                Icon(
                    imageVector = androidx.compose.material.icons.Icons.Default.ChevronRight,
                    contentDescription = null,
                    tint = TextTertiary,
                    modifier = Modifier.size(DesignSystem.IconSize.small)
                )
            }
            
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.sm)
            ) {
                if (category != null) {
                    Surface(
                        shape = RoundedCornerShape(DesignSystem.CornerRadius.round),
                        color = Primary.copy(alpha = 0.1f)
                    ) {
                        Text(
                            text = category.uppercase(),
                            style = MaterialTheme.typography.labelSmall,
                            color = Primary,
                            fontWeight = FontWeight.Medium,
                            modifier = Modifier.padding(
                                horizontal = DesignSystem.Spacing.sm,
                                vertical = DesignSystem.Spacing.xs
                            )
                        )
                    }
                }
                
                if (distance != null) {
                    Text(
                        text = distance,
                        style = MaterialTheme.typography.labelMedium,
                        color = TextSecondary
                    )
                }
            }
        }
    }
}

@Composable
fun NotificationCard(
    title: String,
    message: String,
    timestamp: String,
    isRead: Boolean,
    actions: List<NotificationAction>,
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    BaseCard(
        modifier = modifier.clickable { onClick() },
        backgroundColor = if (isRead) Card else Surface,
        elevation = if (isRead) DesignSystem.Elevation.small else DesignSystem.Elevation.medium
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.sm)
        ) {
            Row(
                verticalAlignment = Alignment.Top,
                horizontalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.md)
            ) {
                Column(
                    modifier = Modifier.weight(1f),
                    verticalArrangement = Arrangement.spacedBy(2.dp)
                ) {
                    Text(
                        text = title,
                        style = MaterialTheme.typography.titleMedium,
                        color = TextPrimary,
                        fontWeight = if (isRead) FontWeight.Normal else FontWeight.SemiBold,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis
                    )
                    
                    Text(
                        text = message,
                        style = MaterialTheme.typography.bodyMedium,
                        color = TextSecondary,
                        maxLines = 3,
                        overflow = TextOverflow.Ellipsis
                    )
                }
                
                if (!isRead) {
                    Box(
                        modifier = Modifier
                            .size(8.dp)
                            .clip(CircleShape)
                            .background(Primary)
                    )
                }
            }
            
            Text(
                text = timestamp,
                style = MaterialTheme.typography.labelMedium,
                color = TextTertiary
            )
            
            if (actions.isNotEmpty()) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.sm)
                ) {
                    actions.forEach { action ->
                        OutlinedButton(
                            onClick = action.onClick,
                            colors = ButtonDefaults.outlinedButtonColors(
                                contentColor = action.color
                            ),
                            border = ButtonDefaults.outlinedButtonBorder.copy(
                                brush = androidx.compose.ui.graphics.SolidColor(action.color)
                            ),
                            shape = RoundedCornerShape(DesignSystem.CornerRadius.sm),
                            modifier = Modifier.height(32.dp)
                        ) {
                            Text(
                                text = action.title,
                                style = MaterialTheme.typography.labelMedium,
                                fontWeight = FontWeight.Medium
                            )
                        }
                    }
                }
            }
        }
    }
}

@Composable
fun EmptyStateCard(
    icon: String,
    title: String,
    message: String,
    actionTitle: String?,
    onActionClick: (() -> Unit)?,
    modifier: Modifier = Modifier
) {
    BaseCard(
        modifier = modifier,
        backgroundColor = Surface
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.lg)
        ) {
            Icon(
                imageVector = androidx.compose.material.icons.Icons.Default.List,
                contentDescription = null,
                tint = TextTertiary,
                modifier = Modifier.size(DesignSystem.IconSize.xxl)
            )
            
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.sm)
            ) {
                Text(
                    text = title,
                    style = MaterialTheme.typography.headlineSmall,
                    color = TextPrimary,
                    textAlign = androidx.compose.ui.text.style.TextAlign.Center
                )
                
                Text(
                    text = message,
                    style = MaterialTheme.typography.bodyMedium,
                    color = TextSecondary,
                    textAlign = androidx.compose.ui.text.style.TextAlign.Center
                )
            }
            
            if (actionTitle != null && onActionClick != null) {
                PrimaryButton(
                    text = actionTitle,
                    onClick = onActionClick
                )
            }
        }
    }
}

enum class TaskStatus(val color: Color) {
    Active(Primary),
    Completed(Success),
    Muted(TextTertiary)
}

data class NotificationAction(
    val title: String,
    val color: Color,
    val onClick: () -> Unit
)

@Preview(showBackground = true)
@Composable
fun CardComponentsPreview() {
    NearMeTheme {
        Column(
            modifier = Modifier.padding(DesignSystem.Spacing.md),
            verticalArrangement = Arrangement.spacedBy(DesignSystem.Spacing.md)
        ) {
            TaskCard(
                title = "Buy groceries",
                description = "Pick up milk, bread, and eggs from the store",
                location = "Whole Foods Market",
                status = TaskStatus.Active,
                onClick = {}
            )
            
            PlaceCard(
                name = "Central Park",
                address = "New York, NY 10024",
                category = "Park",
                distance = "0.5 mi",
                onClick = {}
            )
            
            NotificationCard(
                title = "Task Reminder",
                message = "You're near Whole Foods Market. Don't forget to buy groceries!",
                timestamp = "5 minutes ago",
                isRead = false,
                actions = listOf(
                    NotificationAction("Complete", Success) {},
                    NotificationAction("Snooze", TextSecondary) {}
                ),
                onClick = {}
            )
            
            EmptyStateCard(
                icon = "list",
                title = "No Tasks Yet",
                message = "Create your first task to get started with location-based reminders.",
                actionTitle = "Create Task",
                onActionClick = {}
            )
        }
    }
}
